File: rust/avp-goose/src/runstate.rs
//! Per-run state and turn buffering.
//!
//! An assistant turn accumulates its content blocks and the events it spawns
//! (tool invocations, returns, subagent frames); `drain` closes the turn by
//! emitting a single `assistant_message` (with computed cost) followed by those
//! buffered events in arrival order. This preserves the AVP ordering invariant:
//! the model's message lands before the tool activity it triggered.
//!
//! Goose yields a complete `Message` per assistant turn (not streamed chunks),
//! so a turn maps to one `AgentEvent::Message`. The emitter (separate layer)
//! decides when to open a turn, feeds content/usage/events here, and calls
//! `drain` at the turn boundary.

use std::time::Instant;

use avp::ids::new_span_id;
use avp::pricing::{compute_cost, CostSource, PriceTable};
use avp::sink::Sink;
use avp::trajectory::{AvpContentItem, Usage};
use avp::Event;

use crate::events;

fn zero_usage() -> Usage {
    Usage {
        input_tokens: 0,
        output_tokens: 0,
        cache_read_input_tokens: None,
        cache_creation_input_tokens: None,
        reasoning_output_tokens: None,
    }
}

/// Buffered state for one open assistant turn.
struct Turn {
    step: u64,
    span_id: String,
    started_at: Instant,
    content: Vec<AvpContentItem>,
    usage: Usage,
    response_model: Option<String>,
    /// Events spawned during the turn (tool_invoked / tool_returned /
    /// subagent_*), flushed after the turn's `assistant_message`.
    emissions: Vec<Event>,
}

/// Flat per-run state. One instance per agent run.
pub struct RunState<S: Sink> {
    /// Supervisor-issued run id; CloudEvents `subject` on every event.
    pub run_id: String,
    /// OTel trace id stamped on every event's `data`.
    pub trace_id: String,
    /// Root agent span; parent of every turn / tool / subagent frame.
    pub agent_span_id: String,
    /// Provider name for `assistant_message` (e.g. "anthropic"); `None` if unknown.
    pub provider: Option<String>,
    prices: PriceTable,
    /// 1-based turn counter.
    step: u64,
    turn: Option<Turn>,
    sink: S,
}

impl<S: Sink> RunState<S> {
    pub fn new(
        sink: S,
        run_id: impl Into<String>,
        trace_id: impl Into<String>,
        agent_span_id: impl Into<String>,
        provider: Option<String>,
        prices: PriceTable,
    ) -> Self {
        Self {
            run_id: run_id.into(),
            trace_id: trace_id.into(),
            agent_span_id: agent_span_id.into(),
            provider,
            prices,
            step: 0,
            turn: None,
            sink,
        }
    }

    /// The open turn, opening a fresh one (next step, new span) on first use.
    fn turn_mut(&mut self) -> &mut Turn {
        if self.turn.is_none() {
            self.step += 1;
            self.turn = Some(Turn {
                step: self.step,
                span_id: new_span_id(),
                started_at: Instant::now(),
                content: Vec::new(),
                usage: zero_usage(),
                response_model: None,
                emissions: Vec::new(),
            });
        }
        self.turn.as_mut().expect("just opened")
    }

    /// Open (or reuse) the current turn, returning its `(step, span_id)`. The
    /// emitter uses the span as the `parent_span_id` for the turn's tool /
    /// subagent events.
    pub fn open_turn(&mut self) -> (u64, String) {
        let turn = self.turn_mut();
        (turn.step, turn.span_id.clone())
    }

    /// Emit an event straight through the sink, bypassing the turn buffer. For
    /// lifecycle events and for tool returns that arrive after their turn has
    /// already drained.
    pub fn emit(&self, event: &Event) -> std::io::Result<()> {
        self.sink.emit(event)
    }

    /// Append an assistant content block to the open turn.
    pub fn push_content(&mut self, block: AvpContentItem) {
        self.turn_mut().content.push(block);
    }

    /// Set the open turn's usage and response model (last write wins).
    pub fn set_usage(&mut self, usage: Usage, response_model: Option<String>) {
        let turn = self.turn_mut();
        turn.usage = usage;
        turn.response_model = response_model;
    }

    /// Buffer an event spawned during the turn; flushed after the turn's
    /// `assistant_message` on `drain`.
    pub fn buffer_event(&mut self, event: Event) {
        self.turn_mut().emissions.push(event);
    }

    /// Close the open turn: emit one `assistant_message`, then every buffered
    /// event in arrival order. No-op when no turn is open.
    pub fn drain(&mut self) -> std::io::Result<()> {
        let Some(turn) = self.turn.take() else {
            return Ok(());
        };
        let duration_ms = turn.started_at.elapsed().as_millis() as u64;
        let model = turn.response_model.unwrap_or_default();
        // A turn with no observed usage reports `unknown` rather than a
        // misleading `computed $0` (e.g. an intermediate turn whose tokens
        // Goose has not yet flushed to the session — see TECH_DEBT).
        let (cost_usd, cost_source) =
            if turn.usage.input_tokens == 0 && turn.usage.output_tokens == 0 {
                (0.0, CostSource::Unknown)
            } else {
                compute_cost(
                    &model,
                    turn.usage.input_tokens,
                    turn.usage.output_tokens,
                    turn.usage.cache_read_input_tokens.unwrap_or(0),
                    turn.usage.cache_creation_input_tokens.unwrap_or(0),
                    &self.prices,
                )
            };
        let message = events::assistant_message(
            &self.run_id,
            &self.trace_id,
            &turn.span_id,
            &self.agent_span_id,
            turn.step,
            duration_ms,
            turn.content,
            turn.usage,
            cost_usd,
            cost_source,
            self.provider.as_deref(),
            (!model.is_empty()).then_some(model.as_str()),
        );
        self.sink.emit(&message)?;
        for event in turn.emissions {
            self.sink.emit(&event)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testkit::CapturingSink;
    use avp::pricing::load_default_prices;
    use avp::trajectory::{AvpContentItem, TextBlock};

    fn text(s: &str) -> AvpContentItem {
        AvpContentItem::TextBlock(TextBlock {
            text: s.to_string(),
            citations: None,
            type_: "text".to_string(),
        })
    }

    fn usage(input: u64, output: u64, cache_read: u64, cache_write: u64) -> Usage {
        Usage {
            input_tokens: input,
            output_tokens: output,
            cache_read_input_tokens: Some(cache_read),
            cache_creation_input_tokens: Some(cache_write),
            reasoning_output_tokens: None,
        }
    }

    fn run(sink: CapturingSink) -> RunState<CapturingSink> {
        RunState::new(
            sink,
            "r1",
            "0".repeat(32),
            "1".repeat(16),
            Some("anthropic".to_string()),
            load_default_prices(),
        )
    }

    #[test]
    fn drain_emits_assistant_message_then_buffered_events_in_order() {
        let sink = CapturingSink::default();
        let mut rs = run(sink.clone());

        rs.push_content(text("hello"));
        rs.set_usage(
            usage(1000, 500, 200, 100),
            Some("claude-opus-4-7".to_string()),
        );
        // A buffered tool-ish event (any event; we assert ordering, not shape).
        let buffered: Event = serde_json::from_str(
            r#"{"specversion":"1.0","id":"x","time":"2026-01-01T00:00:00Z","subject":"r1",
                "type":"avp.agent_stopped","source":"avp://agent",
                "data":{"trace_id":"00000000000000000000000000000000",
                "span_id":"2222222222222222","parent_span_id":"0000000000000000",
                "avp.reason":"converged"}}"#,
        )
        .unwrap();
        rs.buffer_event(buffered);

        rs.drain().unwrap();

        let evs = sink.events.lock().unwrap();
        assert_eq!(evs.len(), 2, "assistant_message + 1 buffered event");
        // Assistant message comes first.
        assert_eq!(evs[0]["type"], "avp.assistant_message");
        assert_eq!(evs[0]["data"]["avp.step"], 1);
        assert_eq!(evs[0]["data"]["avp.content"][0]["text"], "hello");
        assert_eq!(evs[0]["data"]["avp.cost.source"], "computed");
        // fresh=700 -> 700*5 + 200*0.5 + 100*6.25 + 500*25 = 16725 (per 1e6).
        let cost = evs[0]["data"]["avp.cost_usd"].as_f64().unwrap();
        assert!((cost - 0.016725).abs() < 1e-9, "got {cost}");
        assert_eq!(evs[0]["data"]["avp.response.model"], "claude-opus-4-7");
        // Buffered event flushed after.
        assert_eq!(evs[1]["type"], "avp.agent_stopped");
    }

    #[test]
    fn drain_clears_turn_and_increments_step() {
        let sink = CapturingSink::default();
        let mut rs = run(sink.clone());

        rs.push_content(text("one"));
        rs.set_usage(usage(10, 5, 0, 0), Some("claude-opus-4-7".to_string()));
        rs.drain().unwrap();

        // Second drain with no open turn is a no-op.
        rs.drain().unwrap();
        assert_eq!(sink.events.lock().unwrap().len(), 1);

        // Next turn gets step 2 and a distinct span.
        rs.push_content(text("two"));
        rs.drain().unwrap();
        let evs = sink.events.lock().unwrap();
        assert_eq!(evs.len(), 2);
        assert_eq!(evs[1]["data"]["avp.step"], 2);
        assert_ne!(evs[0]["data"]["span_id"], evs[1]["data"]["span_id"]);
        // Turn span sits under the agent span.
        assert_eq!(evs[1]["data"]["parent_span_id"], "1".repeat(16));
    }

    #[test]
    fn unknown_model_costs_zero_unknown() {
        let sink = CapturingSink::default();
        let mut rs = run(sink.clone());
        rs.push_content(text("x"));
        rs.set_usage(usage(100, 50, 0, 0), Some("mystery/model".to_string()));
        rs.drain().unwrap();
        let evs = sink.events.lock().unwrap();
        assert_eq!(evs[0]["data"]["avp.cost_usd"], 0.0);
        assert_eq!(evs[0]["data"]["avp.cost.source"], "unknown");
    }
}

File: agents/avp-claude-agent-sdk/python/src/avp_claude_agent_sdk/_runstate.py
"""Per-run state for AVP trajectory emission."""

from __future__ import annotations

import contextvars
import dataclasses
import time
from collections.abc import AsyncIterable
from typing import Any

from avp.content import AVPContentBlock
from avp.envelope import new_span_id
from avp.pricing import ModelPrice, compute_cost
from avp.sink import EventSink
from avp.trajectory import (
    AssistantMessageData,
    AssistantMessageEvent,
    Event,
    Usage,
)

_PROVIDER_NAME = "anthropic"


@dataclasses.dataclass
class TaskInfo:
    """Bookkeeping for a dispatched Task (subagent).

    Populated by `_on_task_started` when the SDK fires `TaskStartedMessage`;
    consumed by `_on_task_notification` to emit the matching
    `subagent_returned`. `span_id` is shared between the invoked and
    returned events (same frame, opened then closed), while
    `parent_span_id` (the turn's span) is what both events sit under.
    """

    # Subagent frame's span; reused by `subagent_returned` (same frame).
    span_id: str
    # Enclosing turn span; parent of both subagent_invoked and subagent_returned.
    parent_span_id: str
    # Turn step the dispatch belongs to.
    step: int
    # Subagent type / name (e.g. "general-purpose"); from `TaskStartedMessage.task_type`.
    task_type: str
    # Monotonic clock at dispatch; drives `subagent_returned.duration_ms`.
    started_at: float


@dataclasses.dataclass
class ToolSpan:
    """Bookkeeping for an in-flight tool dispatch.

    Populated by `avp_pretooluse_hook` on dispatch; popped by
    `avp_posttooluse_hook` on return to assemble the matching
    `tool_returned` (parent under `span_id`, reuse `step` and `name`,
    derive `duration_ms` from `started_at`).
    """

    # `tool_invoked.span_id`; becomes the return event's `parent_span_id`.
    span_id: str
    # Turn step the invocation belongs to; copied onto the return event.
    step: int
    # Tool name; the SDK's `ToolResultBlock` doesn't carry it back.
    name: str
    # Monotonic clock at dispatch; drives `tool_returned.duration_ms`.
    started_at: float


@dataclasses.dataclass
class Turn:
    """Buffered state for one open inference (one Anthropic `message_id`).

    Events produced inside the turn (tool_invoked, subagent_invoked, ...)
    are appended to `emissions` and flushed in arrival order after the
    turn's `assistant_message`. Captured timestamps on each event
    preserve original occurrence times across the deferred flush.
    """

    # Anthropic API response id; one inference per id.
    message_id: str | None
    # 1-based turn counter within the run.
    step: int
    # Monotonic clock at open; used for `duration_ms`.
    started_at: float = dataclasses.field(default_factory=time.monotonic)
    # Turn's own span identity; `parent_span_id` for buffered emissions.
    span_id: str = dataclasses.field(default_factory=new_span_id)
    # Model name from the first chunk (`AssistantMessage.model`).
    response_model: str | None = None
    # API response totals; chunks duplicate the same numbers so
    # last-write-wins converges to the inference's true total.
    usage: Usage = dataclasses.field(default_factory=lambda: Usage(input_tokens=0, output_tokens=0))
    # Last-seen `AssistantMessage.stop_reason` across the chunks.
    stop_reason: str | None = None
    # AVP blocks merged from each chunk in arrival order.
    content: list[AVPContentBlock] = dataclasses.field(default_factory=list)
    # Events captured during the turn; flushed after `assistant_message`.
    emissions: list[Event] = dataclasses.field(default_factory=list)
    # Live tool dispatches keyed by `tool_use_id`. Populated by the
    # PreToolUse hook, popped by the PostToolUse hook to pair returns.
    tool_spans: dict[str, ToolSpan] = dataclasses.field(default_factory=dict)
    # Anthropic service tier from `usage.service_tier`; affects pricing.
    meta_service_tier: str | None = None
    # Cache-creation tokens split by TTL bucket; Anthropic prices these differently.
    meta_cache_creation_5m: int = 0
    meta_cache_creation_1h: int = 0
    # Diagnostic: how many SDK AssistantMessage chunks merged into this turn.
    meta_chunks_merged: int = 0
    # Set once a tool-result for this turn arrives: the inference has ended, so
    # the next AssistantMessage opens a new turn even if it shares (or lacks) a
    # `message_id`. Drain is still deferred to that next message so parallel
    # tool results (which may span several UserMessages) all land on this turn.
    tool_resulted: bool = False


def _turn_has_output(turn: Turn) -> bool:
    """Whether a turn produced anything worth emitting: any buffered event
    (tool_invoked / subagent_*), or any content block other than an empty
    text block. A turn with no content (or only empty text) is degenerate
    and skipped."""
    if turn.emissions:
        return True
    return any(
        not (getattr(b, "type", None) == "text" and not getattr(b, "text", ""))
        for b in turn.content
    )


@dataclasses.dataclass
class RunState:
    """Flat per-run state. One instance per active run."""

    # Prompt the agent was started with
    prompt: str | AsyncIterable[dict[str, Any]] | None
    # Where every event for this run is written. Transport is the caller's choice.
    sink: EventSink
    # Supervisor-issued run identifier; CloudEvents `subject` on every event.
    run_id: str
    # OTel-style 16-byte hex trace id stamped on every event's `data`.
    trace_id: str
    # Per-model price catalog consumed by `compute_cost` at turn drain.
    prices: dict[str, ModelPrice]
    # Commission `enabled_builtin_tools` allow-list (None = expose all). Applied
    # to `agent_started.data.avp.tools` so the merged surface reflects the
    # Commission's subtractive filter over the descriptor's tool bag.
    enabled_builtin_tools: list[str] | None = None
    # Root agent span; parent of every turn / tool / subagent frame.
    # `None` until `agent_started` fires.
    agent_span_id: str | None = None
    # Currently open turn buffer; `None` between turns.
    turn: Turn | None = None
    # Live Task (subagent) dispatches keyed by `tool_use_id`. Populated by
    # `_on_task_started`, popped by `_on_task_notification`.
    #
    # Run-scoped, not turn-scoped. The `Agent` tool dispatches asynchronously:
    # its `tool_result` is a launch receipt that returns in milliseconds, and
    # the child's `TaskNotificationMessage` can arrive many turns later or not
    # at all. Keyed on the turn, the pairing entry died at turn rollover and
    # the frame was never closed. `TaskInfo` captures its own span / parent /
    # step at dispatch, so nothing here depends on the dispatching turn still
    # being open.
    tasks: dict[str, TaskInfo] = dataclasses.field(default_factory=dict)
    # Highest step actually emitted; the next turn opens at `last_step + 1`.
    # Tracked on the run (not derived from the just-drained turn) so a gated
    # empty turn doesn't consume a step number.
    last_step: int = 0
    # `True` once `agent_stopped` has fired; guards against double-emit.
    stopped: bool = False

    async def drain(self) -> Turn | None:
        """Close the open turn: emit one `assistant_message`, then flush
        every buffered event in arrival order. No-op when no turn is open,
        and a no-emit no-op when the turn produced nothing (an empty-output
        inference is not put on the wire, and does not consume a step).
        Clears `self.turn` so the next AssistantMessage opens a fresh one.
        Returns the just-drained `Turn` (or `None` if nothing was emitted).
        """
        if self.turn is None:
            return None
        if self.agent_span_id is None:
            raise RuntimeError("cannot drain turn before agent_started has fired")

        turn = self.turn
        self.turn = None

        if not _turn_has_output(turn):
            return None
        self.last_step = turn.step

        duration_ms = int((time.monotonic() - turn.started_at) * 1000)
        cost_usd, cost_source = compute_cost(
            turn.response_model or "",
            provider=_PROVIDER_NAME,
            input_tokens=turn.usage.input_tokens,
            output_tokens=turn.usage.output_tokens,
            cache_read=turn.usage.cache_read_input_tokens or 0,
            cache_write=turn.usage.cache_creation_input_tokens or 0,
            prices=self.prices,
        )
        meta: dict[str, Any] = {
            "anthropic.message_id": turn.message_id,
            "anthropic.service_tier": turn.meta_service_tier,
            "anthropic.cache_creation.ephemeral_5m_input_tokens": turn.meta_cache_creation_5m,
            "anthropic.cache_creation.ephemeral_1h_input_tokens": turn.meta_cache_creation_1h,
            "claude_agent_sdk.chunks_merged": turn.meta_chunks_merged,
        }
        await self.sink(
            AssistantMessageEvent(
                subject=self.run_id,
                data=AssistantMessageData(
                    trace_id=self.trace_id,
                    span_id=turn.span_id,
                    parent_span_id=self.agent_span_id,
                    meta=meta,
                    step=turn.step,
                    duration_ms=duration_ms,
                    content=turn.content,
                    provider_name=_PROVIDER_NAME,
                    request_model=turn.response_model,
                    response_model=turn.response_model,
                    response_finish_reasons=([turn.stop_reason] if turn.stop_reason else None),
                    usage=turn.usage,
                    cost_usd=cost_usd,
                    cost_source=cost_source,
                ),
            )
        )
        for event in turn.emissions:
            await self.sink(event)

        return turn


# ---------------------------------------------------------------------------
# Ambient run access (asyncio-task-scoped via contextvars)
# ---------------------------------------------------------------------------

_current: contextvars.ContextVar[RunState | None] = contextvars.ContextVar(
    "avp_claude_agent_run", default=None
)


def current_run() -> RunState | None:
    """Read the active `RunState` for this asyncio task / thread, or `None`."""
    return _current.get()


def set_run(state: RunState) -> contextvars.Token[RunState | None]:
    """Bind `state` as the active run; pair with `reset_run(token)` on cleanup."""
    return _current.set(state)


def reset_run(token: contextvars.Token[RunState | None]) -> None:
    """Restore the prior active run using the token returned by `set_run`."""
    _current.reset(token)
